drivers/gpio/gpio_davinci.c

```diff
@@ -0,0 +1,189 @@
+#include <zephyr/devicetree.h>
+
+#include <zephyr/drivers/gpio/gpio_utils.h>
+
+/* Helper Macros for GPIO */
+#define DEV_GPIO_CFG(dev)	\
+		((const struct gpio_davinci_config * const)((dev)->config))
+#define DEV_GPIO_CFG_BASE(dev) \
+		((struct gpio_davinci_regs *)(DEV_GPIO_CFG(dev)->port_base))
+
+#define GPIO_DAVINCI_DIR_RESET_VAL	(0xFFFFFFFF)
+
+struct gpio_davinci_regs {
+	uint32_t dir;
+	uint32_t out_data;
+	uint32_t set_data;
+	uint32_t clr_data;
+	uint32_t in_data;
+	uint32_t set_ris_trig;
+	uint32_t clr_ris_tirg;
+	uint32_t set_fal_trig;
+	uint32_t clr_fal_trig;
+	uint32_t intstat;
+};
+
+struct gpio_davinci_data {
+	struct gpio_driver_data common;
+	sys_slist_t cb;
+};
+
+struct gpio_davinci_config {
+	void (*bank_config)(const struct device *dev);
+	struct gpio_driver_config common;
+	uint32_t port_base;
+	uint32_t port_num;
+};
+
+static int gpio_davinci_configure(const struct device *dev, gpio_pin_t pin,
+					gpio_flags_t flags)
+{
+	volatile struct gpio_davinci_regs *regs = DEV_GPIO_CFG_BASE(dev);
+
+	if ((flags & GPIO_SINGLE_ENDED) != 0) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) != 0) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & GPIO_OUTPUT) != 0) {
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			regs->set_data = BIT(pin);
+		} else {
+			regs->clr_data = BIT(pin);
+		}
+
+		regs->dir &= (~(BIT(pin)));
+
+	} else {
+		regs->dir |= (BIT(pin));
+	}
+
+	return 0;
+}
+
+static int gpio_davinci_port_get_raw(const struct device *dev,
+					gpio_port_value_t *value)
+{
+	volatile struct gpio_davinci_regs *regs = DEV_GPIO_CFG_BASE(dev);
+
+	*value = regs->in_data;
+
+	return 0;
+}
+
+static int gpio_davinci_port_set_masked_raw(const struct device *dev,
+		gpio_port_pins_t mask, gpio_port_value_t value)
+{
+	volatile struct gpio_davinci_regs *regs = DEV_GPIO_CFG_BASE(dev);
+
+	regs->out_data = (regs->out_data & (~mask)) | (mask & value);
+
+	return 0;
+}
+
+static int gpio_davinci_port_set_bits_raw(const struct device *dev,
+						gpio_port_pins_t mask)
+{
+	volatile struct gpio_davinci_regs *regs = DEV_GPIO_CFG_BASE(dev);
+
+	regs->set_data |= mask;
+
+	return 0;
+}
+
+static int gpio_davinci_port_clear_bits_raw(const struct device *dev,
+						gpio_port_pins_t mask)
+{
+	volatile struct gpio_davinci_regs *regs = DEV_GPIO_CFG_BASE(dev);
+
+	regs->clr_data |= mask;
+
+	return 0;
+}
+
+static int gpio_davinci_port_toggle_bits(const struct device *dev,
+						gpio_port_pins_t mask)
+{
+	volatile struct gpio_davinci_regs *regs = DEV_GPIO_CFG_BASE(dev);
+
+	regs->out_data ^= mask;
+
+	return 0;
+}
+
+static const struct gpio_driver_api gpio_davinci_driver_api = {
+	.pin_configure = gpio_davinci_configure,
+	.port_get_raw = gpio_davinci_port_get_raw,
+	.port_set_masked_raw = gpio_davinci_port_set_masked_raw,
+	.port_set_bits_raw = gpio_davinci_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_davinci_port_clear_bits_raw,
+	.port_toggle_bits = gpio_davinci_port_toggle_bits
```

Comment:
Minor nit,
Can we not reuse the kernel naming convention just for maintaining similarity?
Ie. davinci_gpio_xxx
https://elixir.bootlin.com/linux/v6.5-rc4/source/drivers/gpio/gpio-davinci.c#L145

Reply:
In zephyr, all gpio driver functions are defined as gpio_<module_name>.